### ISSUE TITLE
Export including deleted transactions

### ISSUE THREAD

**Issue by mark-adigital** [enupal/stripe#180]
Hi,

There's a couple of different possible areas to address here, but they were all connected as we saw them so hopefully you don't mind me covering the whole thing in one issue. So to try and provide context, our implementation of this plugin is used in tandem with an event listener that adds some CMS records based on the `afterProcessWebhook` event. We had an issue where the module listening for this webhook had some issues which we patched.

The first thing this highlighted was that as the webhook was ultimately returning an error, Stripe was re-sending it at various intervals resulting in duplicate records of the payment in the CMS at `enupal-stripe/orders`, so I'm thinking it may be worth adding some check for whether that Stripe reference is already in the database before creating another new payment record?

Secondly, it appears that when exporting results from `enupal-stripe/orders` using the export button above the graph, the deleted Stripe Payments are still included (this doesn't occur when using the Craft-native export button at the bottom of the table however). I may be incorrect here, but I wonder if this is ultimately due to the soft-delete function now in Craft?

Thanks,

Mark

**Comment by andrelopez** [comment]
Hi @mark-adigital good catch about the webhook, I'll research about that and will keep you updated also about the export issue (you're right this is due to the new soft delete)

**Comment by aarmitage** [comment]
Hi @andrelopez - this webhook error is starting to create some pretty major problems for us with a client and we're going to have to demonstrate that this is being recognised and there's a solution in play. The issue was first reported in March and it took until May to get a response and there's no indication of when a further update will be provided. 

I appreciate the issue has been labelled as an enhancement rather than a bug, but if this is due to the soft delete function in Craft, then clearly the plugin isn't able to keep pace with Craft which makes it unreliable for us. As a paid plugin with which you offer updates and support, I hoped some direction forward with this would have been provided by now.

What are the options for us please?

If this requires a piece of paid work to get it resolved then please let us know. We can't be left hanging with open issues that leave us abandoned impacting on our own reputation and that of Craft itself.

**Comment by andrelopez** [comment]
@aarmitage @mark-adigital Could you please give me more info on how to recreate issue number 1? what is the event type triggered on the re-sending behavior? How can I recreate the error (re-try behavior) ?

**Comment by mark-adigital** [comment]
Hi @andrelopez  - so we have a module listening for the `Orders::EVENT_AFTER_ORDER_COMPLETE` event, that was sometimes failing due to the constraints on the entry record it was creating, like missing required fields (though I don't think the particular reason for the failure makes a difference). The core issue is that if an event listener fails, it the webhook returns a 500 error to Stripe, which then re-sends the webhook at various intervals. So instead of the Payments plugin always creating a new order in Craft, I was thinking it might make sense to first query the Stripe `craft_enupalstripe_orders` table to see whether an order with the matching Stripe Transaction ID already exists and only create the order if required?

**Comment by andrelopez** [comment]
@mark-adigital Thanks for following up. Actually, we already check if the order exists, take a look to this line:

https://github.com/enupal/stripe/blob/master/src/controllers/WebhookController.php#L35

As you can see we try to get the order from the next transaction id `$eventJson['data']['object']['id']`

I think what you can do is create an Order Status for example 'Error' and instead of Raise errors on your listener just set the order status to 'Error' you can add messages to your order so you can add more info about the cause of the error:
```php
use enupal\stripe\Stripe;
...
# you can pass an object or an array to the last argument and it will encode it to JSON
Stripe::$app->messages->addMessage($order->id, 'Error from my listener module', $object);
```
Please let me know if that makes sense

**Comment by mark-adigital** [comment]
Thanks for this @andrelopez .

If I'm honest, I think it sort of makes sense! Based on what you're saying I'm wondering if part of the issue is that the orders in question aren't for physical products but for event entries, meaning there's no order processing and the order status never changes from 'New'. Could that be why the orders aren't being detected despite already existing?

I don't believe the event listener code that was throwing the 500 error is doing anything with the order itself so am not sure what difference updating the status to be an error would make. Would it be worth me sending the event listener code across to you can see what it's doing?

**Comment by mark-adigital** [comment]
I don't know if it helps, but the errors we're seeing are generally timeouts now (which tend to succeed on the second attempt) and despite the check in the code, we're still seeing duplicate orders with the same Stripe Transaction ID are created in the control panel, and therefore multiple order emails are also being sent to customers.

**Comment by andrelopez** [comment]
Hi @mark-adigital we released Stripe Payments 3.3.10 to fix the deleted transactions issue, please feel free to open a new issue related to the other issue you mentioned

**Comment by andrelopez** [comment]
Hi @mark-adigital We just released Stripe Payments 3.3.11 with a fix to the duplicate orders issue (if for some reason Stripe triggers the checkout event multiple times)
